jax_galsim/spergel.py

```diff
@@ -0,0 +1,424 @@
+import galsim as _galsim
+import jax
+import jax.numpy as jnp
+import tensorflow_probability as tfp
+from jax._src.numpy.util import _wraps
+from jax.tree_util import Partial as partial
+from jax.tree_util import register_pytree_node_class
+
+from jax_galsim.core.draw import draw_by_kValue, draw_by_xValue
+from jax_galsim.core.utils import bisect_for_root, ensure_hashable
+from jax_galsim.gsobject import GSObject
+from jax_galsim.random import UniformDeviate
+
+
+@jax.jit
+def _Knu(nu, x):
+    """Modified Bessel 2nd kind"""
+    return tfp.substrates.jax.math.bessel_kve(nu * 1.0, x) / jnp.exp(jnp.abs(x))
+
+
+@jax.jit
+def _gamma(nu):
+    """Gamma(nu) with care for integer nu in [0,5]"""
+    return jnp.select(
+        [nu == 0, nu == 1, nu == 2, nu == 3, nu == 4, nu == 5],
+        [jnp.inf, 1.0, 1.0, 2.0, 6.0, 24.0],
+        default=jnp.exp(jax.lax.lgamma(nu * 1.0)),
+    )
+
+
+@jax.jit
+def _gammap1(nu):
+    """Gamma(nu+1)"""
+    return _gamma(nu + 1.0)
+
+
+@jax.jit
+def z2lz(z):
+    """return z^2 * log(z)"""
+    return jnp.where(z <= 1e-40, 0.0, z * z * jnp.log(z))
+
+
+@jax.jit
+def f0(z):
+    """K_0[z] z -> 0  O(z^4)"""
+    z2 = z * z
+    z4 = z2 * z2
+    c0 = 0.11593151565841244881
+    c1 = 0.27898287891460311220
+    c2 = 0.025248929932162694513
+    return c0 + c1 * z2 + c2 * z4 - jnp.power(1.0 + 0.125 * z2, 2.0) * jnp.log(z)
+
+
+@jax.jit
+def f1(z):
+    """z^1 K_1[z] z -> 0  O(z^4)"""
+    z2 = z * z
+    z4 = z2 * z2
+    c0 = z2lz(z)  # z^2 log(z)
+    c1 = 0.30796575782920622441
+    c2 = 0.08537071972865077805
+    return 1.0 - c1 * z2 - c2 * z4 + c0 * (0.5 + 0.0625 * z2)
+
+
+@jax.jit
+def f2(z):
+    """z^2 K_2[z] z -> 0  O(z^4)"""
+    c1 = 0.10824143945730155610
+    z2 = z * z
+    z4 = z2 * z2
+    c0 = z2lz(z) * z2  # z^4*log(z)
+    return 2.0 - 0.5 * z2 + c1 * z4 - 0.125 * c0
+
+
+@jax.jit
+def f3(z):
+    """z^3 K_3[z] z -> 0  O(z^4)"""
+    z2 = z * z
+    z4 = z2 * z2
+    return 8.0 - z2 + 0.125 * z4
+
+
+@jax.jit
+def f4(z):
+    """z^4 K_4[z] z -> 0 O(z^4)"""
+    z2 = z * z
+    z4 = z2 * z2
+    return 48.0 - 4 * z2 + 0.25 * z4
+
+
+@jax.jit
+def f5(z):
+    """z^5 K_5[z] z -> 0 O(z^4)"""
+    z2 = z * z
+    z4 = z2 * z2
+    return 384.0 - 24.0 * z2 + z4
+
+
+@jax.jit
+def fsmallz_nu(z, nu):
+    def fnu(z, nu):
+        """z^nu K_nu[z] z -> 0 O(z^4) z > 0"""
+        z2 = z * z
+        z4 = z2 * z2
+        c1 = jnp.power(2.0, -6.0 - nu)
+        c2 = _gamma(-2.0 - nu)
+        c3 = _gamma(-2.0 + nu)
```

Comment:
This variable is not used. Is that a bug?

Reply:
Thanks, I've missed c3 coefficient.